rootston/desktop.c

```diff
@@ -425,12 +432,10 @@ static void view_handle_new_subsurface(struct wl_listener *listener,
 	subsurface_create(view, wlr_subsurface);
 }
 
-void view_init(struct roots_view *view, struct roots_desktop *desktop) {
-	assert(view->wlr_surface);
+void view_map(struct roots_view *view, struct wlr_surface *surface) {
```

Comment:
The interface seems a little bit misleading since you aren't able to unmap a view, then map it with an arbitrary different surface. The view might already be set up with a different type of surface and they won't match.

Reply:
Would it be better to not have the `surface` as a function parameter and just get the surface from the shell surface implementation?

Reply:
>The interface seems a little bit misleading since you aren't able to unmap a view, then map it with an arbitrary different surface.

The map/unmap API is private to shells. Only the shell can decide which wl_surface needs to be mapped.

>Would it be better to not have the surface as a function parameter and just get the surface from the shell surface implementation?

That would mean adding `view->wl_surface = surface` before each `view_map`. Not sure it's better.